File: implement-shell-tools/wc/wc.py
import sys
import os

# Flag Detection
show_lines = "-l" in sys.argv
show_words = "-w" in sys.argv
show_chars = "-c" in sys.argv

# If no flags then show all.
show_all = not (show_lines or show_words or show_chars)

# Filter files excluding flags
files = [arg for arg in sys.argv[1:] if not arg.startswith("-")]

# Total counters
total_l, total_w, total_c = 0, 0, 0

for filename in files:
    try:
        with open(filename, 'r') as f:
            lines = f.readlines()
            
            # files calculation
            l_count = len(lines)
            w_count = sum(len(line.split()) for line in lines)
            #  os.path.getsize for bytes, in .js I used const bytes = Buffer.byteLength(content);
            c_count = os.path.getsize(filename)
            
            # Update totals
            total_l += l_count
            total_w += w_count
            total_c += c_count
            
            # Printing Logic
            output = []
            if show_lines or show_all: output.append(f"{l_count:>8}")
            if show_words or show_all: output.append(f"{w_count:>8}")
            if show_chars or show_all: output.append(f"{c_count:>8}")
            
            print(f"{''.join(output)} {filename}")

    except FileNotFoundError:
        print(f"wc: {filename}: No such file or directory")
    except IsADirectoryError:
        print(f"wc: {filename}: Is a directory")
        print(f"{'0':>8} {'0':>8} {'0':>8} {filename}")

# Print Total if there were multiple files
if len(files) > 1:
    output_total = []
    if show_lines or show_all: output_total.append(f"{total_l:>8}")
    if show_words or show_all: output_total.append(f"{total_w:>8}")
    if show_chars or show_all: output_total.append(f"{total_c:>8}")
    print(f"{''.join(output_total)} total")

File: implement-shell-tools/wc/wc.js
import process, { exit } from "node:process";
import fs from "node:fs";

const argv = process.argv.slice(2);

let showWords = argv.includes("-w");
let showLines = argv.includes("-l");
let showBytes = argv.includes("-c");
let showCharacters = argv.includes("-m");

const filePaths = argv.filter((arg) => !arg.startsWith("-"));
// if no flags enable all.
if (!showLines && !showCharacters && !showWords && !showBytes) {
  showLines = true;
  showCharacters = true;
  showWords = true;
  showBytes = true;
}

if (filePaths.length === 0) {
  console.error("PLease provide a file path");
  process.exit(1);
}

let totalLines = 0;
let totalWords = 0;
let totalBytes = 0;
let totalChars = 0;

filePaths.forEach((filePath) => {
  try {
    const content = fs.readFileSync(filePath, "utf-8");
    const lineArray = content.split("\n");
    if (content.endsWith("\n")) lineArray.pop(); 
    const lines = lineArray.length;
    
    const words = content.trim().split(/\s+/).filter((word) => word != "").length;
    const bytes = Buffer.byteLength(content);
    const characters = content.length;

    totalLines += lines;
    totalWords += words;
    totalBytes += bytes;
    totalChars += characters;

    let output = "";
    if (showLines)   output += lines.toString().padStart(8);
    if (showWords)   output += words.toString().padStart(8);
    if (showBytes)   output += bytes.toString().padStart(8);
    if (showCharacters) output += characters.toString().padStart(8)

    console.log(`${output} ${filePath}`); 
  } catch(error) {
    console.error(`Error reading ${filePath}: ${error.message}`);
  }
}); 

if (filePaths.length > 1) {
  let totalOutput = "";
  if (showLines)  totalOutput += `${totalLines.toString().padStart(8)}`;
  if (showWords)  totalOutput += `${totalWords.toString().padStart(8)}`;
  if (showBytes)  totalOutput += `${totalBytes.toString().padStart(8)}`;
  if (showCharacters) totalOutput += `${totalChars.toString().padStart(8)}`;

  console.log(`${totalOutput} total`);
}
  
